storage/range.go

```diff
@@ -760,42 +759,79 @@ func (r *Range) proposeRaftCommand(ctx context.Context, args proto.Request, repl
 // struct to get args and reply and then applying the command to the
 // state machine via applyRaftCommand(). The error result is sent on
 // the command's done channel, if available.
-//
-// TODO(spencer): Differentiate between errors caused by the normal culprits --
-// bad inputs from clients, stale information, etc. and errors which might
-// cause the range replicas to diverge -- running out of disk space, underlying
-// rocksdb corruption, etc. Do a careful code audit to make sure we identify
-// errors which should be classified as a ReplicaCorruptionError--when those
-// bubble up to the point where we've just tried to execute a Raft command, the
-// Raft replica would need to stall itself.
 func (r *Range) processRaftCommand(idKey cmdIDKey, index uint64, raftCmd proto.InternalRaftCommand) error {
 	if index == 0 {
 		log.Fatalc(r.context(), "processRaftCommand requires a non-zero index")
 	}
+
 	r.Lock()
 	cmd := r.pendingCmds[idKey]
 	delete(r.pendingCmds, idKey)
 	r.Unlock()
 
 	args := raftCmd.Cmd.GetValue().(proto.Request)
+	origin := proto.RaftNodeID(raftCmd.OriginNodeID)
+	var reply proto.Response
+	var ctx context.Context
+	var err error
 	if cmd != nil {
 		// We initiated this command, so use the caller-supplied reply.
-		err := r.applyRaftCommand(cmd.ctx, index, proto.RaftNodeID(raftCmd.OriginNodeID), args, cmd.Reply)
-		cmd.done <- err
-		return err
+		reply = cmd.Reply
+		ctx = cmd.ctx
+	} else {
+		// This command originated elsewhere so we must create a new reply buffer.
+		reply = args.CreateReply()
+		ctx = r.context()
+	}
+
+	// Verify the leader lease is held; Note that we don't require the leader
+	// lease when trying to grant the leader lease!
+	held, expired := verifyLeaderLease(r.getLease(), origin, args.Header().Timestamp)
+	if (!held || expired) && args.Method() != proto.InternalLeaderLease {
+		// This error case is somewhat special: Any command that makes it into
+		// Raft should have the leader lease held by the proposer. The only
+		// case that violates this assumption is that of the proposer having a
+		// corrupted internal state and erroneously assuming leadership. Then,
+		// the command should certainly not be executed by the replicas, but
+		// the error must not enter the response cache (which is updated in
+		// applyRaftCommand below) since it is going to be retried with exactly
+		// the same timestamp. On retry, it may be proposed by the real leader
+		// (or a later leader, which thereby assumes leadership over all past
+		// timestamps as well), for which the command should execute. If it did
+		// hit the response cache, the distributed sender would get stuck in an
+		// infinite loop, retrieving a stale NotLeaderError over and over
+		// again.
+		err = r.newNotLeaderError()
+	} else {
+		err = r.maybeSetCorrupt(
+			r.applyRaftCommand(ctx, index, origin, args, reply),
```

Comment:
It's a little weird to see applyRaftCommand (the central action of this function) buried as a parameter to maybeSetCorrupt. I'd pull this out into separate statements (and maybe change `maybeSetCorrupt` to `if err != nil { err = r.setCorrupt() }`)